### PR TITLE
[Color+Elevation] Add a convenience method to resolve dynamic color only when traitCollection's color appearance has changed.

### PR DESCRIPTION
**Pull request by fluxminds**
Add this convenience method would help supporting theming extension in an easier way that we don't need to check color difference on traitCollection every time like https://github.com/material-components/material-components-ios/pull/8285/files#diff-48ca693ee2870d02b06a42900a1dc69dR73.

### CODE REVIEW

File: components/Elevation/src/UIColor+MaterialElevation.h [hunk 1]
```diff
@@ -30,6 +30,23 @@
  */
 - (nonnull UIColor *)mdc_resolvedColorWithElevation:(CGFloat)elevation;
 
+/**
+ Returns a color that takes the specified elevation value and traits into account when there is a
+ color appearance difference between current traits and previous traits. When userInterfaceStyle is
+ UIUserInterfaceStyleDark in currentTraitCollection, elevation will be used to resolve the color.
+ Negative elevation is treated as 0.
```
Comment (codeman7) [inline components/Elevation/src/UIColor+MaterialElevation.h:38]: Should we separate these on new lines to make it a little more readable?

File: components/Elevation/src/UIColor+MaterialElevation.h [hunk 2]
```diff
@@ -30,6 +30,23 @@
...
+ Pattern-based UIColor is not supported.
+ UIColor in UIExtendedGrayColorSpace will be resolved to UIExtendedSRGBColorSpace.
+
+ @param currentTraitCollection The traits to use when resolving the color.
+ @param previousTraitCollection The previous traits to use when comparing color appearance.
+ @param elevation The @c mdc_absoluteElevation to use when resolving the color.
+ */
+- (nonnull UIColor *)
+    mdc_resolvedColorWithTraitCollection:(nonnull UITraitCollection *)currentTraitCollection
```
Comment (codeman7) [inline components/Elevation/src/UIColor+MaterialElevation.h]: Can we make this `withTraitCollection:(UITraitCollection *)traitCollection` to match the existing API and UIKit?

File: components/Elevation/tests/unit/MaterialElevationColorTests.m [hunk 3]
```diff
@@ -296,4 +296,50 @@ - (void)testSmoothJumpBetweenElevationToAlphaForValuesCloseToOne {
...
+        [dynamicColor mdc_resolvedColorWithTraitCollection:currentTraitCollection
+                                   previousTraitCollection:previousTraitCollection
+                                                 elevation:elevation];
+    // Then
+    XCTAssertEqualObjects(resolvedColor, dynamicColor);
+  }
+}
+
+- (void)testResolvingColorWithDifferenceCurrentTraitCollectionAndPreviousTraitCollection {
+  if (@available(iOS 13.0, *)) {
```
Comment (codeman7) [inline components/Elevation/tests/unit/MaterialElevationColorTests.m:324]: Should this have the `if defined`?

File: components/Elevation/tests/unit/MaterialElevationColorTests.m [hunk 4]
```diff
@@ -296,4 +296,50 @@ - (void)testSmoothJumpBetweenElevationToAlphaForValuesCloseToOne {
...
+    UITraitCollection *previousTraitCollection =
+        [UITraitCollection traitCollectionWithUserInterfaceStyle:UIUserInterfaceStyleDark];
+    UITraitCollection *currentTraitCollection =
+        [UITraitCollection traitCollectionWithUserInterfaceStyle:UIUserInterfaceStyleLight];
+    UIColor *resolvedColor =
+        [dynamicColor mdc_resolvedColorWithTraitCollection:currentTraitCollection
+                                   previousTraitCollection:previousTraitCollection
+                                                 elevation:elevation];
+    // Then
+    XCTAssertNotEqual(resolvedColor, dynamicColor);
```
Comment (codeman7) [inline components/Elevation/tests/unit/MaterialElevationColorTests.m:342]: Can we test that the trait collection = what UIKit would return?

**Review by codeman7 — APPROVED** [review]

File: components/Elevation/src/UIColor+MaterialElevation.h (hunk 1 above)
Comment (fluxminds) [reply components/Elevation/src/UIColor+MaterialElevation.h:38]: Addressed in 75d9d11dfb3fd470687a570978baba739a411c59.

File: components/Elevation/src/UIColor+MaterialElevation.h (hunk 2 above)
Comment (fluxminds) [reply components/Elevation/src/UIColor+MaterialElevation.h]: Addressed in 75d9d11dfb3fd470687a570978baba739a411c59.

File: components/Elevation/tests/unit/MaterialElevationColorTests.m (hunk 3 above)
Comment (fluxminds) [reply components/Elevation/tests/unit/MaterialElevationColorTests.m:324]: Good point! Addressed in 75d9d11dfb3fd470687a570978baba739a411c59.

File: components/Elevation/tests/unit/MaterialElevationColorTests.m (hunk 4 above)
Comment (fluxminds) [reply components/Elevation/tests/unit/MaterialElevationColorTests.m:342]: Addressed in 75d9d11dfb3fd470687a570978baba739a411c59.

**Review by yarneo — APPROVED** [review]
nice!

**Pull request merged**